File: mobile/lib/service/greeter_service.dart
import 'package:makeaplan/api/greeter_service.pb.dart' as pb;
import 'package:makeaplan/api/greeter_service.pbgrpc.dart' as grpc;
import 'package:makeaplan/model/user_profile.dart';

import 'grpc_client.dart';

class GreeterService {
  static GreeterService _singleton;

  GreeterService._internal();

  factory GreeterService() {
    if (_singleton == null) {
      _singleton = GreeterService._internal();
    }
    return _singleton;
  }

  grpc.GreeterClient _client() {
    return grpc.GreeterClient(
      GRPCClient().client(),
      options: getCommonCallOptions(),
    );
  }

  Future<bool> sayHello() async {
    print("GreeterService.sayHello");
    try {
      await _client().sayHello(pb.HelloRequest.create());
      return true;
    } catch (err) {
      if (err.message == "who is this") {
        return false;
      } else {
        throw err;
      }
    }
  }

  Future<UserProfile> getUserProfile() async {
    print("GreeterService.getUserProfile");
    final v = await _client().sayHello(pb.HelloRequest.create());
    final sps = v.message.split(", ");
    final username = sps[1];
    final dataID = sps[2];
    UserProfile up = UserProfile();
    if (username != "***") {
      up.username = username;
    }
    if (dataID != "***") {
      up.dataID = dataID;
    }

    return up;
  }

  Future<void> start() async {
    print("GreeterService.start");
    await _client().start(pb.StartRequest.create());
  }

  Future<void> updateUsername(String un) async {
    print("GreeterService.updateUsername");
    await _client()
        .updateUsername(pb.UpdateUsernameRequest.create()..username = un);
  }

  Future<void> updateDeviceID(String did) async {
    print("GreeterService.updateDeviceID");
    await Future.delayed(Duration(seconds: 2));
    await _client()
        .updateDeviceID(pb.UpdateDeviceIDRequest.create()..dataId = did);
  }

  Future<void> initPerSetPlan(int id) async {
    print("GreeterService.initPerSetPlan");
    await _client().initPerSetPlan(pb.InitPerSetPlanRequest.create()..id = id);
  }
}

File: service/makeaplan/service/greeter_service.go
package service

import (
	"context"
	"errors"
	"fmt"

	"github.com/guregu/null"
	"google.golang.org/grpc"

	"makeaplan.com/api/service"
	"makeaplan.com/app"
	"makeaplan.com/dao"
	"makeaplan.com/model"
)

type greeterService struct {
}

func RegisterGreeterService(server *grpc.Server) {
	service.RegisterGreeterService(server, service.NewGreeterService(&greeterService{}))
}

func (g *greeterService) UpdateUsername(ctx context.Context, request *service.UpdateUsernameRequest) (*service.UpdateUsernameReply, error) {
	uid := app.UIDFromContext(ctx)
	mp := &model.User{}
	mp.Username = null.StringFrom(request.GetUsername())
	_, _, err := dao.UpdateUser(ctx, uid, mp)
	if err != nil {
		return nil, err
	}

	return &service.UpdateUsernameReply{}, nil
}

func (g *greeterService) UpdateDeviceID(ctx context.Context, request *service.UpdateDeviceIDRequest) (*service.UpdateDeviceIDReply, error) {
	did := app.DIDFromContext(ctx)
	if reachQuota(ctx, did, request.GetDataId()) {
		app.Printf(ctx, "reached. to_did: %s, usedDataID: %s\n", did, request.GetDataId())
		return nil, errors.New("reach quota")
	}

	user, err := dao.FindUserByDataID(ctx, request.GetDataId())
	if err != nil {
		return nil, err
	}
	oldUser, err := dao.FindUser(ctx, did)
	if err != nil && err != dao.ErrNotFound {
		return nil, err
	}
	if err != dao.ErrNotFound && oldUser != nil {
		if oldUser.DataId == request.GetDataId() {
			return &service.UpdateDeviceIDReply{}, nil
		}

		oldUser.DeviceId = fmt.Sprintf("take by user with id: %d", user.ID)
		_, _, err = dao.UpdateUser(ctx, oldUser.ID, oldUser)
		if err != nil {
			return nil, err
		}
	}

	mp := &model.User{}
	mp.DeviceId = did
	_, _, err = dao.UpdateUser(ctx, user.ID, mp)
	if err != nil {
		return nil, err
	}

	return &service.UpdateDeviceIDReply{}, nil
}

func (g *greeterService) Start(ctx context.Context, request *service.StartRequest) (*service.StartReply, error) {
	did := app.DIDFromContext(ctx)
	dtid, err := dao.GenerateUserDataID(ctx)
	if err != nil {
		return nil, err
	}

	mu := &model.User{
		DeviceId:   did,
		UserStatus: model.UserStatusNormal,
		DataId:     dtid,
	}
	_, _, err = dao.AddUser(ctx, mu)
	if err != nil {
		return nil, fmt.Errorf("failed to dao.AddUser, err: %v", err)
	}

	return &service.StartReply{}, nil
}

func (g *greeterService) SayHello(ctx context.Context, request *service.HelloRequest) (*service.HelloReply, error) {
	uid := app.UIDFromContext(ctx)
	user, err := dao.GetUser(ctx, uid)
	if err != nil {
		return nil, err
	}

	un := "***"
	if user.Username.Valid {
		un = user.Username.String
	}
	return &service.HelloReply{
		Message: fmt.Sprintf("Hello there, %s, %s", un, user.DataId),
	}, nil
}

func (g *greeterService) InitPerSetPlan(ctx context.Context, request *service.InitPerSetPlanRequest) (*service.InitPerSetPlanReply, error) {
	uid := app.UIDFromContext(ctx)
	err := dao.InsertPreSetPlan(request.GetId(), uid)
	if err != nil {
		return nil, err
	}

	return &service.InitPerSetPlanReply{}, nil
}
